nx/lib/nx/binary_backend.ex

```diff
@@ -1571,6 +1571,89 @@ defmodule Nx.BinaryBackend do
     end
   end
 
+  @impl true
+  def scatter_add(
+        %T{} = out,
+        %T{shape: shape, type: {_, target_size}} = target,
+        %T{shape: {indices_rows, _indices_cols} = indices_shape} = indices,
+        %T{shape: {indices_rows}} = updates,
+        _opts
+      ) do
+    indices_bin_list =
+      indices |> to_binary() |> aggregate_axes([1], indices_shape, elem(indices.type, 1))
+
+    target_byte_size = div(target_size, 8)
+
+    offsets_list =
+      match_types [indices.type] do
+        for idx_bin <- indices_bin_list do
+          idx = for <<match!(x, 0) <- idx_bin>>, do: read!(x, 0)
+
+          {offset, []} =
+            idx
+            |> Enum.with_index()
+            |> Enum.reduce(
+              {0, Tuple.to_list(shape)},
+              fn {idx, axis}, {offset, [dim_size | shape]} ->
+                if idx < 0 or idx >= dim_size do
+                  raise ArgumentError,
+                        "index #{idx} is out of bounds for axis #{axis} in shape #{inspect(shape)}"
+                end
+
+                {offset + idx * Enum.product(shape), shape}
+              end
+            )
+
+          offset * target_byte_size
+        end
+      end
+
+    updates_list =
+      match_types [updates.type] do
+        for <<match!(x, 0) <- to_binary(updates)>>, do: read!(x, 0)
+      end
+
+    {offsets_with_updates, _last_offset} =
+      offsets_list
+      |> Enum.zip(updates_list)
+      |> Enum.group_by(fn {off, _} -> off end, fn {_, upd} -> upd end)
+      |> Enum.sort_by(fn {off, _} -> off end)
+      |> Enum.map_reduce(0, fn {next_offset, upds}, previous_offset ->
+        {{
+           previous_offset + target_byte_size,
+           next_offset,
+           Enum.sum(upds)
+         }, next_offset}
+      end)
+
+    target_binary = to_binary(target)
+
+    {result, tail} =
+      for {previous, current, update} <- offsets_with_updates, reduce: {<<>>, target_binary} do
+        {traversed, to_traverse} ->
+          before_slice_size = max(current - previous, 0)
+          before_offset = binary_part(to_traverse, 0, before_slice_size)
+
+          element = binary_part(to_traverse, before_slice_size, target_byte_size)
+
+          total_size = byte_size(to_traverse)
+
+          to_traverse =
+            binary_part(
+              to_traverse,
+              before_slice_size + target_byte_size,
+              total_size - (before_slice_size + target_byte_size)
+            )
+
+          updated_element =
+            scalar_to_binary(binary_to_number(element, target.type) + update, out.type)
+
+          {traversed <> before_offset <> updated_element, to_traverse}
+      end
```

Comment:
I think this overall function looks solid but this last part of updating the binary seems a bit expensive. I will write pseudo-codeish but hopefully it shows you how I would tackle it using recursion:

    def scatter_add([{previous, current, update} | rest], binary) do
      match_types [type] do
        <<previous_bin::size(previous), match!(x, 0), rest_bin::binary>> = binary
        <<previous_bin::size(current - previous), write!(read!(x, 0) + update, 0), scatter_add(rest, rest_bin)::binary>>
      end
    end

    def scatter_add([], binary), do: binary

I think the compiler will be smart enough to optimize it too. :)

Reply:
Oh, I see! Maybe I could just accumulate an iodata list and the iodata_to_binary in the end? This would change almost nothing in my code.

Reply:
@polvalente yes, that would also work! You would need a revert in your code and in mine it could be list recursive, but it all turns out to be the same.

Reply:
Wouldn't `[acc | [iter_result]]` bypass the Enum.reverse?

Reply:
true!!!